DRILL-4729: Add support for prepared statement implementation on server side

vkorukanti opened the pull request:
+ Add following APIs for Drill Java client
  - DrillRpcFuture<CreatePreparedStatementResp> createPreparedStatement(final String query)
  - void executePreparedStatement(final PreparedStatement preparedStatement, UserResultsListener resultsListener)
  - List<QueryDataBatch> executePreparedStatement(final PreparedStatement preparedStatement) (for testing purpose)

+ Separated out the interface from UserClientConnection. It makes it easy to have wrappers which need to
  tap the messages and data going to the actual client.

+ Implement CREATE_PREPARED_STATEMENT and handle RunQuery with PreparedStatement

+ Test changes to support prepared statement as query type

+ Add tests in TestPreparedStatementProvider

parthchandra commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/prepare/PreparedStatementProvider.java:
```diff
@@ -0,0 +1,419 @@
+public class PreparedStatementProvider {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(PreparedStatementProvider.class);
+
+  private static final int LIMIT_0_QUERY_TIMEOUT_MINS = 10;
```
I would make this limit configurable. Ideally, the client should be able to cancel the prepare statement request. There is no way to get the query id for the prepare (i.e limit0) query, is there?

vkorukanti replied:
10mins is already a large number for `LIMIT 0` query timeout. I don't mind adding another option, but we already have too many options. It is possible the query could timeout for scheduling reasons, in which case we can send the status request as `TIMEOUT` (currently we have `SUCCESS`, `FAILURE`). Also we may not even need a limit 0 query in future if we optimize this path, so sending the `LIMIT 0` query id to client is a maintenance overhead later on.

Looking at the JDBC and ODBC documentation, it doesn't look like there is a timeout for create prepare statement (correct me if I am wrong here). There is only timeout for execute prepare statement.

parthchandra replied:
I agree, 10 min is a long time. Your call on whether to make it configurable, but given our current implementation, it is not impossible for us to hit it.
For the Prepare call, a cancel is probably a good idea in the wire protocol; that's just standard behaviour for any async protocol. But that may be a future change.

sudheeshkatkam replied:
Or make this a system property?

We have seen cases where `LIMIT 0` queries take longer than 10 minutes. Although there is [an option (disabled by default)](https://issues.apache.org/jira/browse/DRILL-3623) to avoid full-fledged execution for queries on _tables with known schema_, on other tables, `LIMIT 0` queries may take 10 minutes. They should not, but they may (due to single fragment execution).

vkorukanti replied:
I added a session/system option in latest commit. In my opinion if a `LIMIT 0` query is taking longer than 10mins, we should try to fix it, otherwise it is not an interactive experience when using Drill with BI tools.

vkorukanti commented:
Added option for changing the timeout. Also added a new RequestStatus type `TIMEOUT`. @parthchandra New changes are in a separate commit. I will create a JIRA for adding cancel.

sudheeshkatkam commented on exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java:
```diff
@@ -512,6 +526,61 @@ public void runQuery(QueryType type, List<PlanFragment> planFragments, UserResul
+   * @return List of {@link QueryDataBatch}s. It is responsibility of the caller to release query data batches.
+   * @throws RpcException
+   */
+  public List<QueryDataBatch> executePreparedStatement(final PreparedStatementHandle preparedStatementHandle)
```
VisibleForTesting annotation?

vkorukanti replied:
done

sudheeshkatkam commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/prepare/PreparedStatementProvider.java:
```diff
@@ -0,0 +1,427 @@
+    private final CountDownLatch latch = new CountDownLatch(1);
+
+    private DrillPBError error;
+    private List<SerializedField> fields;
```
volatile

vkorukanti replied:
thanks for catching this.

sudheeshkatkam commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/prepare/PreparedStatementProvider.java:
```diff
@@ -0,0 +1,427 @@
+    return builder.build();
+  }
+
+  private static int getPrecision(MajorType majorType) {
```
Move these helpers to `Types` class?

vkorukanti replied:
moved to Types

sudheeshkatkam commented:
+1, pending minor comments

sudheeshkatkam commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/prepare/PreparedStatementProvider.java:
```diff
@@ -0,0 +1,427 @@
+
+        final QueryId limit0QueryId = userWorker.submitWork(wrapper, limit0Query);
+
+        final long timeout_millis =
```
timeoutMillis

vkorukanti replied:
done

vkorukanti commented:
pushed the squashed commit that contains the review changes.
